### ISSUE TITLE
Replace `Swarm.Language.Key` with `Brick.Keybindings.Pretty`

### ISSUE THREAD

**Issue by xsebek** [swarm-game/swarm#1978]
#### Is your feature request related to a problem? Please describe.
Since Brick has gotten custom keybindings support, it also has functions for parsing and pretty-printing keybindings.

It would be best to switch to the Brick implementation which would remove some of our code and make bugs less likely.

#### Describe the solution you'd like
Replace functions in `Swarm.Language.Key` with functions from:
* [`Brick.Keybindings.Pretty`](https://hackage.haskell.org/package/brick-1.3/docs/Brick-Keybindings-Pretty.html)
* [`Brick.Keybindings.Parse`](https://hackage.haskell.org/package/brick-1.3/docs/Brick-Keybindings-Parse.html)

#### Describe alternatives you've considered

We could keep our own implementation, but given it depends on `Vty`, there is currently no benefit to it.

If we also reimplemented `Vty` parts, then we could at least have `swarm-lang` not depend on Brick.

#### Additional context
I noticed this while hacking on #523.

**Comment by byorgey** [comment]
Makes sense. In theory, if the text format `brick` uses for keybindings is different than ours, this could break some existing Swarm programs, but I don't really care.

Very happy to hear you are hacking on #523. :grin:

**Comment by kostmo** [comment]
> Conceptually it would also be very nice if `swarm-lang` did not depend on `brick` at all.

I would like this as well - this would make the "core" more portable.  One difficulty in getting the tournament server working (although I've worked around it for now) was the dependency on `vty` that was not supported on a certain build target ([reference](https://github.com/swarm-game/swarm/blob/40159975adb460e275193cd5b463255fdb7cc151/tournament/scripts/docker/build-server-executable.sh#L8-L9)).  I wonder how palatable it might be upstream to break the [`Key` datatype](https://hackage.haskell.org/package/vty-5.37/docs/Graphics-Vty-Input-Events.html#t:Key) into a separate package?

**Comment by xsebek** [comment]
@kostmo I suppose that would be best if `vty`/`brick` is open to the idea.

Alternatively, we could have our own `Key` type and a bijection to `vty` keys. We could then pass the brick parsing and pretty printing functions to `GameState` from `TUI` as a dependency injection. For unit testing purposes, we could pass functions that fail if used.

**Comment by kostmo** [comment]
> Alternatively, we could have our own `Key` type and a bijection to `vty` keys.

@jtdaugherty, we would appreciate your thoughts on how best to obtain a lightweight dependency on key definitions.

**Comment by jtdaugherty** [comment]
> @jtdaugherty, we would appreciate your thoughts on how best to obtain a lightweight dependency on keyboard key definitions.

Reading over this, I'm at a bit of a loss. Some thoughts:

Given that `swarm-lang` already implements parsing and pretty-printing of key information in what is presumably your desired format, I would recommend sticking with that; the version provided by `brick` is provided as a convenience. Using `brick`'s version might allow you to remove some code, but it's true that that would make your backend library dependent on `brick`. I don't see how that's a big problem if you are wanting `brick` to be authoritative for how key information is represented, but if you don't want that, then it would make sense to me to roll your own (which you already have at the moment).

It sounds like the main motivation here is conceptual, which I can appreciate; I would imagine that you want your backend library not to depend on things that are more client-specific. Setting aside the possibility of depending on `brick` for a moment, assuming `swarm-lang` depends only on `vty`, I guess I don't understand what makes that burdensome. (@kostmo mentioned an issue with `tinfo` when building `vty` in Amazon docker containers, but it sounds like that was worked around using `cabal`, and it sounded like that wasn't a problem.) It's true that, *conceptually*, depending on `vty` might be strange in `swarm-lang` if `swarm-lang` has nothing to do with clients - but in a way, it does anyway, if you are wanting to use `vty`'s data types for keys and modifiers there. It would seem that you've already committed to that view of input, so depending on `vty` in other places in your infrastructure doesn't strike me as a problem. If that *is* a problem, then perhaps you do need your own abstract representation for input anyway, if (say) you want to relegate all input handling details to other libraries, thus leaving open the possibility of SDL or web clients.

Assuming a "lightweight dependency" here means "don't make me build all of `vty`", I don't have any solutions readily at hand. I certainly don't want to break just the input types out into their own package; divorced from this particular situation, that would be strange, and I suspect such a package would have exactly one user. It would introduce some extra maintenance overhead for me, with no conceptual gain for `vty`. In the mean time, I guess I am not finding a compelling argument here that depending on `vty` itself is heavyweight, ignoring the `tinfo` issue since it can be worked around.

I do want to help if I can, but I don't see a path forward here other than committing to depending on `vty` (which seems reasonable to me because you've already committed to it in `swarm-lang` if you want its key and modifier types) or making your own internal key/modifier types and translating those to Vty, which would not be very nice and would amount to duplicating code.

**Comment by byorgey** [comment]
@jtdaugherty Thanks for the thoughts!

**Comment by byorgey** [comment]
I think I'm going to close this, since I'm not convinced it is desirable.  Feel free to reopen if there are more concrete thoughts on why we would want this and what a good path forward would be.
